File: Data Structures & Algorithms/search-2d-matrix/submission-8.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int row_size = matrix.size(), col_size = matrix[0].size();
        int left = 0, right = row_size;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (matrix[mid][0] < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        if (left < row_size && matrix[left][0] == target) return true;
        if (left == 0) return false;
        int target_row = left - 1;
        left = 0, right = col_size;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (matrix[target_row][mid] < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        if (left < col_size && matrix[target_row][left] == target) return true;
        return false;
    }
};

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-8.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode[]} lists
     * @return {ListNode}
     */
    conquer(list1, list2) {
        let dummy = new ListNode(), node = dummy;
        while (list1 && list2) {
            if (list1.val <= list2.val) {
                node.next = list1;
                list1 = list1.next;
            } else {
                node.next = list2;
                list2 = list2.next;
            }
            node = node.next;
        }
        if (list1) node.next = list1;
        else node.next = list2;
        return dummy.next;
    }

    divide(lists, l, r) {
        if (l > r) return null;
        if (l == r) return lists[l];
        const mid = Math.floor(l + (r - l) / 2);  // avoid overflow
        const left = this.divide(lists, l, mid);
        const right = this.divide(lists, mid + 1, r);
        return this.conquer(left, right);
    }

    mergeKLists(lists) {
        if (!lists) return nullptr;
        return this.divide(lists, 0, lists.length - 1);
    }
}
